### ISSUE TITLE
amphp/dns/lib/DefaultResolver.php(632): Undefined offset: x

### ISSUE THREAD

**Issue by vanoostrum** [amphp/dns#64]
There seems to be a bug in amphp/dns v0.8.15. I'm getting Undefined offset errors in `DefaultResolver`

/amphp/dns/lib/DefaultResolver.php line 632, method `processDecodedResponse`:

```
private function processDecodedResponse($serverId, $requestId, $response) {
        list($promisor, $name, $type, $uri, $timeout) = $this->pendingRequests[$requestId];
        //...
}
```

`$this->pendingRequests` doesn't contain a value at index `$requestId`

Stacktrace:

```
#1 vendor/amphp/dns/lib/DefaultResolver.php(632): Xel\Cake\Error\AppConsoleErrorHandler->Xel\Cake\Error\{closure}(8, Undefined offset: 13, vendor/amphp/dns/lib/DefaultResolver.php, 632, array)
#2 vendor/amphp/dns/lib/DefaultResolver.php(622): Amp\Dns\DefaultResolver->processDecodedResponse(557, 13, LibDNS\Messages\Message)
#3 vendor/amphp/dns/lib/DefaultResolver.php(597): Amp\Dns\DefaultResolver->decodeResponsePacket(string, string)
#4 vendor/amphp/amp/lib/NativeReactor.php(265): Amp\Dns\DefaultResolver->onReadable(0000000032606eb2000000000d168e0f, resource, null)
#5 vendor/amphp/amp/lib/NativeReactor.php(249): Amp\NativeReactor->doIoCallback(0000000032606eb2000000000d168e0f, stdClass, resource)
#6 vendor/amphp/amp/lib/NativeReactor.php(214): Amp\NativeReactor->selectActionableStreams(0)
#7 vendor/amphp/amp/lib/NativeReactor.php(71): Amp\NativeReactor->doTick(false)
#8 vendor/amphp/amp/lib/functions.php(46): Amp\NativeReactor->run(Closure)
```

The call is coming from `DefaultResolver->query()`, which is called in our case from the library kelunik/acme `Dns01->verifyChallenge()`

**Comment by kelunik** [comment]
Please upgrade to 0.9.x.

**Comment by kelunik** [comment]
This should actually be fixed now, but the root cause isn't fixed and there's probably one.

**Comment by vanoostrum** [comment]
Thanks, I will test it with the new version

**Comment by vanoostrum** [comment]
@kelunik I'm having some dependency issues with some of your other libraries, basically kelunik/acme 0.4.x -> amphp/artax ^2 -> amphp/socket ^0.9 -> amphp/dns ^0.8. Could that be resolved so I can use amphp/dns 0.9 in combination with kelunik/acme 0.4.x?

**Comment by kelunik** [comment]
`kelunik/acme` needs a rewrite based on Amp v2 for that. I started doing that locally, but currently don't have the time to finish it. I could probably just port it to Amp v2 and tag that as 0.5.0 instead of doing a rewrite at the same time.

**Comment by vanoostrum** [comment]
Ok, We are currently using react/dns as an alternative, whenever we can we will switch back to amphp/dns

**Comment by kelunik** [comment]
Which a custom event loop adapter or how do you do that?

**Comment by vanoostrum** [comment]
Something like that yes, while keeping the exact method signature of amphp/dns verifyChallenge

```
use Amp\Deferred;
use Amp\Promise;
use Exception;
use React\Dns\Model\Message;
use React\Dns\Protocol\BinaryDumper;
use React\Dns\Protocol\Parser;
use React\Dns\Query\Executor;
use React\Dns\Query\Query;
use React\EventLoop\Factory;

class DnsResolver {
    public function verifyChallenge($hostName, $text): Promise {
        $deferred = new Deferred();
        $name = "_acme-challenge." . $hostName;
        $loop = Factory::create();
        $executor = new Executor($loop, new Parser(), new BinaryDumper());
        $nameServer = 1.2.3.4 //own nameserver
        $executor->query($nameServer, new Query($name, Message::TYPE_TXT, Message::CLASS_IN, time()))
            ->done(function (Message $message) use ($name, $text, $deferred) {
                if (strpos($message->data, $text) > 0) {
                    //found the correct txt record
                    $deferred->succeed();
                } else {
                    $deferred->fail(new Exception("dns txt record $text not found for $name"));
                }
            }, function(Exception $e) use ($deferred) {
                $deferred->fail($e);
            });

        $loop->run();

        return $deferred->promise();
    }
}
```

**Comment by kelunik** [comment]
@vanoostrum I was able to allocate some time today. I have a working implementation based on Amp v2 in https://github.com/kelunik/acme/tree/amp2. Please test it, if everything works, I can tag a version based on it. The tests pass at least.

**Comment by vanoostrum** [comment]
That's great, I will have to do some refactoring, will let you know if it works

**Comment by vanoostrum** [comment]
@kelunik I see that the AcmeClient now only wants a PrivateKey as constructor arg, not a private/public KeyPair. How does this work exactly? Did you document this change somewhere? How does Lets Encrypt know about the public key in this new situation?

**Comment by vanoostrum** [comment]
It will take some more time to refactor, hopefully it will be ready somewhere next week.

**Comment by kelunik** [comment]
@vanoostrum Changes aren't documented yet, but most things should actually be internal changes apart from the change to the Amp v2 interfaces.

The public key can always be derived from the private key, it's done [here](https://github.com/kelunik/acme/blob/496c9189f37ab512664cb12dc7b994c1ca6965cb/lib/Crypto/Backend/OpensslBackend.php#L24-L42). The public key was actually totally unused.

If you have further feedback or questions, please open them as issue(s) in `kelunik/acme`, so the discussions can be found by others and do not end up in an unrelated component here.

**Comment by vanoostrum** [comment]
Ok that's clear, will do, thx!

**Comment by vanoostrum** [comment]
@kelunik we tested the new version. Everything seems to be working well, also with Amp v2. Thanks for the help! We will keep an eye out to see when the changes will be merged to master, until then we will keep using the amp2 branch

**Comment by kelunik** [comment]
@vanoostrum Thanks for testing. I don't have the time to provide a proper changelog, but I have just released v0.5.0 for you. https://github.com/kelunik/acme/releases/tag/v0.5.0
